File: snapshots/typescript/agent.ts
import { Runloop } from "@runloop/api-client";
import dotenv from "dotenv";
import OpenAI from "openai";
import type { ChatCompletion, ChatCompletionMessage, ChatCompletionTool } from "openai/resources";

dotenv.config();

const RUNLOOP_API_KEY = process.env.RUNLOOP_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

if (!RUNLOOP_API_KEY) {
  throw new Error("Missing required API key: RUNLOOP_API_KEY");
}

if (!OPENAI_API_KEY) {
  throw new Error("Missing required API key: OPENAI_API_KEY");
}

const runloop = new Runloop({ bearerToken: RUNLOOP_API_KEY });
const openai = new OpenAI({ apiKey: OPENAI_API_KEY });

const MAX_ITERATIONS = 10;
const SYSTEM_PROMPT = `
You are a senior game developer tasked with creating a simple Python console game of Snake in a Runloop Devbox.
Each step, you will modify and test the game. If the test is successful, take a snapshot of the Devbox.
`;

const USER_PROMPT = `
You are tasked with creating a simple Python console game of Snake in a Runloop Devbox.
All files are located and executeed in the /home/user/ directory.
Start by checking for a snapshot of the game by getting the list of available snapshots using getSnapshotsList.
Look for a snapshot whose name starts with "snake_game_typescript. 
If one exists, start a Devbox from the snapshot with the highest version number. Follow these guidelines to complete the task:

- If starting from a snapshot and the devbox has been started run 'snake_game.py' using the command **"python /home/user/snake_game.py"** in the Devbox.  
- If the file runs without errors, **increment the version number and take a snapshot** named "snake_game_typescript_<version+1>". 
- If the file does not exist, write the initial version of the game to a file named "snake_game.py" in the Devbox.
- Continue iterating and making improvements autonomously. Each iteration should enhance gameplay mechanics.  
-  **DO NOT WAIT FOR USER INPUT before making improvements.** Always assume the next logical improvement.  
- Take a snapshot every time the file runs successfully without errors.  
- If the game still has errors and **${MAX_ITERATIONS} iterations** have been reached, shut down the Devbox and end the task.  
`;

const tools = [
  {
    type: "function",
    function: {
      name: "executeShellCommand",
      description: "Run a shell command in the devbox",
      parameters: {
        type: "object",
        properties: {
          command: { type: "string" },
        },
        required: ["command"],
        additionalProperties: false,
      },
      strict: true,
    },
  } as const,
  {
    type: "function",
    function: {
      name: "writeFile",
      description: "Writes a file on the devbox at /home/user/<filename>",
      parameters: {
        type: "object",
        properties: {
          filename: { type: "string" },
          contents: { type: "string" },
        },
        required: ["filename", "contents"],
        additionalProperties: false,
      },
      strict: true,
    },
  } as const,
  {
    type: "function",
    function: {
      name: "createSnapshot",
      description: "Takes a snapshot of the Devbox and saves it",
      parameters: {
        type: "object",
        properties: {
          snapshotName: { type: "string" },
        },
        required: ["snapshotName"],
        additionalProperties: false,
      },
      strict: true,
    },
  } as const,
  {
    type: "function",
    function: {
      name: "getSnapshotsList",
      description: "Gets the list of available snapshots",
      parameters: {
        type: "object",
        properties: {},
        additionalProperties: false,
      },
      additionalProperties: false,
      strict: true,
    },
  } as const,
  {
    type: "function",
    function: {
      name: "shutdownDevbox",
      description: "Shuts down the current devbox",
      parameters: {
        type: "object",
        properties: {},
        additionalProperties: false,
      },
      strict: true,
    },
  } as const,
  {
    type: "function",
    function: {
      name: "createDevbox",
      description:
        "Creates a devbox from a snapshot by providing a snapshotId or a new one if not provided",
      parameters: {
        type: "object",
        properties: {
          snapshotId: { type: "string" },
        },
        required: ["snapshotId"],
        additionalProperties: false,
      },
      strict: true,
    },
  } as const,
];

async function runAgent() {
  let devboxId = "";

  const messageHistory: ChatCompletionMessage[] = [
    { role: "assistant", content: SYSTEM_PROMPT, refusal: null },
    {
      role: "user",
      content: USER_PROMPT,
      refusal: null,
    } as unknown as ChatCompletionMessage,
  ];

  let numIterations = 0;
  let response: ChatCompletion | null = null;

  while (numIterations < MAX_ITERATIONS) {
    try {
      response = await openai.chat.completions.create({
        messages: messageHistory,
        model: "gpt-4-turbo",
        tools: tools,
        tool_choice: "auto",
      });

      const latestMessage = response.choices[0].message;
      messageHistory.push(latestMessage);

      if (!latestMessage.tool_calls) break;

      const toolResponses: ChatCompletionMessage[] = [];

      if (latestMessage.tool_calls.length !== 0) {
        for (const toolCall of latestMessage.tool_calls) {
          const functionArgs = JSON.parse(toolCall.function.arguments);
          let result: string;

          if (
            !devboxId &&
            toolCall.function.name !== "createDevbox" &&
            toolCall.function.name !== "getSnapshotsList"
          ) {
            console.log(
              `No devbox exists, and the agent isn't trying to create or list snapshots. Creating a new devbox... ${toolCall.function.name}`,
            );
            devboxId = await runloop.devboxes.createAndAwaitRunning().then((devox) => devox.id);
          }

          console.log(`Executing tool call: ${toolCall.function.name}`);
          switch (toolCall.function.name) {
            case "executeShellCommand":
              {
                const test = await runloop.devboxes.executeSync(devboxId, {
                  command: functionArgs.command,
                });
                result = `${test.stdout} ${test.stderr}`;
                console.log(result);
              }
              break;
            case "writeFile":
              await runloop.devboxes.writeFileContents(devboxId, {
                file_path: functionArgs.filename,
                contents: functionArgs.contents,
              });
              result = "File written successfully";
              break;
            case "createSnapshot":
              await runloop.devboxes.snapshotDisk(devboxId, {
                name: functionArgs.snapshotName,
              });
              result = `Snapshot signal sent with name ${functionArgs.snapshotName}`;
              break;
            case "getSnapshotsList":
              result = JSON.stringify(await runloop.devboxes.listDiskSnapshots());
              break;
            case "shutdownDevbox":
              console.log("Shutting down the devbox...");
              await runloop.devboxes.shutdown(devboxId);
              result = "Devbox shutdown signal sent.";
              break;
            case "createDevbox":
              console.log("Creating the devbox...");
              result = functionArgs.snapshotId
                ? await runloop.devboxes
                    .createAndAwaitRunning({
                      snapshot_id: functionArgs.snapshotId,
                    })
                    .then((devox) => devox.id)
                : (await runloop.devboxes.createAndAwaitRunning()).id;
              devboxId = result;
              result = devboxId;
              break;
            default:
              throw new Error(`Unknown action: ${functionArgs.action}`);
          }

          toolResponses.push({
            role: "tool",
            tool_call_id: toolCall.id,
            content: result,
          } as unknown as ChatCompletionMessage);
        }
      }

      messageHistory.push(...toolResponses);
    } catch (error) {
      console.error("Error during OpenAI API call:", error);
      break;
    }
    numIterations++;
  }

  if (!response) {
    console.error("Error: No response received from OpenAI API after", numIterations, "iterations");
    return;
  }

  console.log(response.choices[0].message.content);
  console.log(`The total number of iterations is ${numIterations}`);
}

async function main() {
  await runAgent();
}

main().catch(console.error);

File: snapshots/python/agent.py
import logging
import os
from runloop_api_client import Runloop
import ell
from ell import Message
from typing import List

RUNLOOP_API_KEY = os.environ.get("RUNLOOP_API_KEY")
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
DEVBOX_ID = ""

if not RUNLOOP_API_KEY:
    raise ValueError("Missing required API key: RUNLOOP_API_KEY")
if not OPENAI_API_KEY:
    raise ValueError("Missing required API key: OPENAI_API_KEY")

MAX_ITERATIONS = 10

SYSTEM_PROMPT = """
You are a senior game developer tasked with creating a simple Python console game of Snake in a Runloop Devbox.
Each step, you will modify and test the game. If the test is successful, take a snapshot of the Devbox.
""".strip()

USER_PROMPT = """
1. If a snapshot named "snake_game_basic_<version>" exists, start a Devbox from the latest snapshot.  
2. Retrieve and analyze logs and `snake_game.py` for errors or improvements.  
3. Fix logged issues and enhance the game iteratively.  
4. Run `snake_game.py` to test functionality after each change.  
5. If the test passes, take a snapshot named "snake_game_basic_<version+1>".  
6. Continue iterating until the game is working without errors.
""".strip()


def run_agent(runloop: Runloop):
    """Runs the AI agent to develop the game iteratively."""

    @ell.tool()
    def execute_shell_command(command: str):
        """Executes a shell command inside the Devbox."""
        result = runloop.devboxes.execute_sync(DEVBOX_ID, command=command)
        return result.stdout if result.stdout else result.stderr

    @ell.tool()
    def write_file(file_name: str, contents: str):
        """Writes a file to the Devbox."""
        runloop.devboxes.write_file_contents(
            DEVBOX_ID, file_path=file_name, contents=contents
        )
        return

    @ell.tool()
    def create_snapshot(snapshot_name: str):
        """Takes a snapshot of the Devbox and saves it."""
        runloop.devboxes.snapshot_disk(DEVBOX_ID, name=snapshot_name)
        return snapshot_name

    @ell.tool()
    def get_latest_snapshot():
        """Retrieves the latest snapshot with the naming pattern 'snake_game_basic_<version>'."""
        try:
            snapshots = (
                runloop.devboxes.list_disk_snapshots().snapshots
            )  # Get response object

            latest_snapshot = None
            max_version = -1

            for snap in snapshots:
                name = snap.name
                if name.startswith("snake_game_basic_"):
                    version = int(name.split("_")[-1])  # Extract numeric version
                    if version > max_version:
                        max_version = version
                        latest_snapshot = snap.id
            return latest_snapshot

        except Exception:
            return None

    @ell.tool()
    def get_devbox_logs():
        """Retrieve the logs from the current devbox and return them in a JSON-serializable format."""
        try:
            logs_response = runloop.devboxes.logs.list(DEVBOX_ID)
            log_entries = logs_response.logs

            # Convert logs into a JSON-serializable list of dictionaries
            formatted_logs = [
                {
                    "cmd_id": log.cmd_id,
                    "level": log.level,
                    "timestamp_ms": log.timestamp_ms,
                    "shell_name": log.shell_name,
                    "cmd": log.cmd,
                    "message": log.message,
                    "exit_code": log.exit_code,
                    "source": log.source,
                }
                for log in log_entries
            ]

            return formatted_logs

        except Exception:
            return "Error retrieving logs"

    @ell.tool()
    def shutdown_devbox():
        """Shuts down the current devbox."""
        runloop.devboxes.shutdown(id=DEVBOX_ID)
        return f"Devbox {DEVBOX_ID} has been shut down"

    @ell.tool()
    def create_devbox(snapshot_id: str):
        """Creates a devbox from a snapshot if provided, otherwise creates a new devbox."""
        global DEVBOX_ID
        if snapshot_id:
            devbox_id = runloop.devboxes.create(snapshot_id=snapshot_id).id
        else:
            devbox_id = runloop.devboxes.create().id
        DEVBOX_ID = devbox_id
        devbox = runloop.devboxes.await_running(devbox_id)
        return devbox.id

    @ell.complex(
        model="gpt-4-turbo",
        tools=[
            execute_shell_command,
            get_devbox_logs,
            get_latest_snapshot,
            write_file,
            shutdown_devbox,
            create_snapshot,
            create_devbox,
        ],
    )
    def invoke_agent(message_history: List[Message]):
        """Invokes the AI agent to develop and test the game."""
        messages = [ell.system(SYSTEM_PROMPT), ell.user(USER_PROMPT)] + message_history
        return messages

    message_history = []
    result = invoke_agent(message_history)
    num_iterations = 0

    while result.tool_calls or num_iterations < MAX_ITERATIONS:
        message_history.append(result)
        result_message = result.call_tools_and_collect_as_message()
        message_history.append(result_message)
        result = invoke_agent(message_history)
        num_iterations += 1

    print(result.text)


def main():
    logger = logging.getLogger(__name__)
    runloop = Runloop(bearer_token=RUNLOOP_API_KEY)

    logger.info("Running AI agent")
    run_agent(runloop)


if __name__ == "__main__":
    main()
